File: Affine Cipher.py
def coprime(multiplier): #This function determines if the multiplier is coprime with 26 
    if (26<=multiplier):
        return False
    a = multiplier
    b = 26
    while a>0:
        x = b%a
        b = a
        a = x
    if b>1 and a==0:
        return False
    else:
        return True
    
def encrypt(shift,multiplier):
    plainText = input("Enter message to encrypt: ")
    encrypted = ""
    for letter in plainText:
        if letter.isupper():
            tempVar = ord(letter)-ord("A") #Temporary value sets letter between 0 and 25
            encrypted+=chr(ord("A")+((multiplier*tempVar)+shift)%26)
        elif letter.islower():
            tempVar = ord(letter)-ord("a")
            encrypted+=chr(ord("a")+((multiplier*tempVar)+shift)%26)
        else:
            encrypted+=letter #If the character isn't a letter, the character will be added back to the message
    print("Encrypted message:",encrypted)

def decrypt(shift,multiplier): 
    a = modInverse(multiplier)
    plainText = input("Enter message to decrypt: ")
    decrypted = ""
    for letter in plainText:
        if letter.isupper():
            tempVar = ord(letter)-ord("A") 
            decrypted+=chr(ord("A")+(a*(tempVar-shift)%26))
        elif letter.islower():
            tempVar = ord(letter)-ord("a")
            decrypted+=chr(ord("a")+((a*(tempVar-shift)%26)))
        else:
            decrypted+=letter
    print("Decrypted message:",decrypted)

def modInverse(multiplier): #While I understand the concept, I could not apply it to code. This function's algorithm is not mine
    for i in range(1, 26): #Taken from https://www.geeksforgeeks.org/multiplicative-inverse-under-modulo-m/ 
        if (((multiplier % 26) * (i % 26)) % 26 == 1):
            return i
    return -1
def main():
    print("Affine Cipher Program")
    shift = int(input("Enter shift: "))
    multiplier = int(input("Enter multiplier: "))
    isCoprime = coprime(multiplier)
    print(isCoprime)
    if isCoprime==True:
        choice = int(input("Enter 1 to encrypt a message, enter 2 to decrypt a message: "))
        if choice==1:
            encrypt(shift,multiplier)
        if choice==2:
            decrypt(shift,multiplier)
        else:
            print("Invalid Input. Please try again.") #Validates input
    else:
        print("Multiplier must be coprime with 26. Please try again.") #Multiplier must be coprime with 26 to properly encrypt/decrypt
main()

File: Affine Cipher.c
#include <stdio.h>
#include <string.h>

#define SIZE 100

void encrypt(int shift, int multiplier); //Function prototypes
void decrypt(int shift, int multiplier);
int coprime(int multiplier);
int modInverse(int multiplier);

int main(){
    int shift, multiplier, choice, isCoprime;
    printf("Affine Cipher Program\n");
    printf("Enter shift: ");
    scanf("%d",&shift);
    printf("\n");
    printf("Enter multiplier: ");
    scanf("%d",&multiplier);
    printf("\n");
    isCoprime = coprime(multiplier);
    if (isCoprime==1){
        printf("Enter 1 to encrypt a message, enter 2 to decrypt a message: ");
        scanf("%d",&choice);
        printf("\n");
        if (choice==1)
            encrypt(shift, multiplier);
        else if (choice==2)
            decrypt(shift,multiplier);
        else
            printf("Invalid input. Please try again."); //Validates Input
    }
    else{
        printf("Multiplier must be coprime with 26. Please try again."); //Multiplier must be coprime with 26 to properly encrypt/decrypt
    }
    return 0;
}

int coprime(int multiplier){
    int a=multiplier, b=26,x;
    if (26<=multiplier)
        return -1;

    while (a>0){
        x = b % a;
        b = a;
        a = x;
    }
    if (b>1 && a==0)
        return -1;
    else
        return 1;
}

void encrypt(int shift, int multiplier){
    char plainText[SIZE], encrypted[SIZE];
    int i,tempVar;
    printf("Enter message to encrypt: ");
    scanf(" %[^\n]s", plainText); //Takes input with spaces
    printf("\n");
    for(i=0;i<strlen(plainText);i++){
        if((int)(plainText[i])>=65&&(int)(plainText[i])<=90){ //uppercase ASCII values
            tempVar = (int)(plainText[i]) - (int)('A'); //uses ASCII values to set letters to values 0-25
            encrypted[i] = (char)((int)('A')+((multiplier*tempVar)+shift)%26);
        }
        else if((int)(plainText[i])>=97&&(int)(plainText[i])<=122){ //lowercase ASCII values
            tempVar = (int)(plainText[i]) - (int)('a');
            encrypted[i] = (char)((int)('a')+((multiplier*tempVar)+shift)%26);
        }
        else{
            encrypted[i] = plainText[i]; //adds any non-alphabetic characters back into the encrypted message
        }
    }
    printf("Encrypted Message: %s",encrypted);
}

void decrypt(int shift, int multiplier){
    char plainText[SIZE], decrypted[SIZE];
    int i,tempVar,a;
    a = modInverse(multiplier);
    printf("Enter message to decrypt: ");
    scanf(" %[^\n]s", plainText);
    printf("\n");
    for(i=0;i<strlen(plainText);i++){
        if((int)(plainText[i])>=65&&(int)(plainText[i])<=90){
            tempVar = (int)(plainText[i])-(int)('A'); 
            decrypted[i] = (char)((int)('A')+(a*(tempVar-shift+26)%26));
        }
        else if ((int)(plainText[i])>=97&&(int)(plainText[i])<=122){
            tempVar = (int)(plainText[i])-(int)('a'); 
            decrypted[i] = (char)((int)('a')+(a*(tempVar-shift+26)%26));
        }
        else{
            decrypted[i] = plainText[i];
        }
    }
    printf("Decrypted Message: %s",decrypted); 
}

int modInverse(int multiplier){
    for(int i;i<26;i++){  
        if (((multiplier % 26) * (i % 26)) % 26 == 1)
            return i;
    }
    return -1;
}